### ISSUE TITLE
What are the rules/schemas for global model references?

### ISSUE THREAD

**Issue by JoergNeidig** [admin-shell-io/questions-and-answers#103]
In the AAS Spec Part 1 the Section 5.3.10.4 contains constraints and examples for the keys. However, looking at the examples of model references for submodelElements I do not understand the schema for creating the respective keys. Is it possible to provide more detailled information on how to create the references, possibly in XML/JSON?

**Comment by arnoweiss** [comment]
The questions that I have:
1. How to I refer to a Referable that's embedded? Examples include:
   1. `AnnotatedRelationshipElement.annotations`
   2. `Entity.statements`
   3. `Capability.inputVariable`
   4. `Capability.outputVariable`
2. What's the reason that [AASd-125](https://aas-core-works.github.io/aas-core-meta/v3/Reference.html#constraint-AASd-125) ("no keys after the first may be globally identifiable") does not extend to GlobalReferences?

**Comment by StenGruener** [comment]
@BirgitBoss [ToDo] please kindly provide an example with XML and Text-Serialization

**Comment by BirgitBoss** [comment]
> The questions that I have:
> 
>     1. How to I refer to a Referable that's embedded? Examples include:
>        
>        1. `AnnotatedRelationshipElement.annotations`
>        2. `Entity.statements`
>        3. `Capability.inputVariable`
>        4. `Capability.outputVariable`
> 

annotations, inputVariable and outputVariable are not referable, only AnnotatedRelationshipElement, Entity, Capability. Attributes within Referables cannot be referenced in the current specification of the Asset Administration Shell (IDTA-01001-3-0)

Do you have use cases that would require to reference these "embedded" attributes.

This would be a new issue for https://github.com/admin-shell-io/aas-specs to also allow References to non-referable model elements.

Up to now the metamodel allows a key/type "FragmentReference", but this is used for (IDTA-01001-3-0):
`Bookmark or a similar local identifier of a subordinate part of a primary resource`

**Comment by arnoweiss** [comment]
The attributes themselves aren't referable but the content they hold is. All have a range of `SubmodelElement` or `Property` respectively. So when adding a `SubmodelElement` INTO these attributes, they are referable. However, they can't be referred to with the existing `Reference` mechanism. That's how they differ from - say - `AssetAdministrationShell.administration` whose target (`AdministrativeInformation`) doesn't inherit from `Referable`. But the elements in the mentioned list do hold `Referable`s, just like `SubmodelElementCollection.value` does.

I don't have a use-case which is why I'll let this topic rest. The `Reference` mechanism is underspecified in this instance.

**Comment by BirgitBoss** [comment]
> The attributes themselves aren't referable but the content they hold is. All have a range of `SubmodelElement` or `Property` respectively. So when adding a `SubmodelElement` INTO these attributes, they are referable. However, they can't be referred to with the existing `Reference` mechanism. That's how they differ from - say - `AssetAdministrationShell.administration` whose target (`AdministrativeInformation`) doesn't inherit from `Referable`. But the elements in the mentioned list do hold `Referable`s, just like `SubmodelElementCollection.value` does.
> 
> I don't have a use-case which is why I'll let this topic rest. The `Reference` mechanism is underspecified in this instance.

you are right

**Comment by BirgitBoss** [comment]
> 
> ```json
> {
>   "modelType": "ReferenceElement",
>   "idShort": "ExampleReferenceElement",
>   "value": {
>     "keys": [
>       {
>         "type": "Submodel",
>         "value": "https://acplt.org/Test_Submodel"
>       },
>       {
>         "type": "SubmodelElementCollection",
>         "value": "ExampleSubmodelElementCollection"
>       },
>       {
>         "type": "Property",
>         "value": "ExampleProperty"
>       }
>     ],
>     "type": "ModelReference"
>   }
> }
> ```
> 

I would make it more generic and use the examples generated by aas-core: 
https://github.com/admin-shell-io/aas-specs/blob/master/schemas/json/examples/generated/Reference/maximal.json
and
https://github.com/admin-shell-io/aas-specs/blo

There also other examples in https://github.com/admin-shell-io/aas-specs/tree/master/schemas/json/examples/generated/Reference

More generic a model reference looks like this:
```json
"<attribute with Type 'Reference'": {
        "keys": [
          {
            "type": "Submodel",
            "value": "urn:an-example08:f3f73640"
          }
        ],
        "type": "ModelReference"
      },
      "id": "something_142922d6",
      "modelType": "<Class of the attribute having the attribute>"
    }
```

**Comment by BirgitBoss** [comment]
But the question is how e.g. the JSON or XML representation is linked to the text serialization as defined as a grammar in cahpter 7.2.3 in IDTA-01001-3-0

Several examples are given there: e.g.
`(Submodel)https://example.com/aas/1/1/1234859590, (SubmodelElementList)Documents, (SubmodelElementCollection)0, (MultiLanguageProperty)Title`

This is identical to the JSON representation:
```json
"<attribute with Type 'Reference'": {
        "keys": [
          {
            "type": "Submodel",
            "value": "https://example.com/aas/1/1/1234859590"
          }
          {
            "type": "SubmodelElementList",
            "value": "Documents"
          }
          {
            "type": "SubmodelElementCollection",
            "value": "0"
          }
          {
            "type": "MultiLanguageProperty",
            "value": "Title"
          }
        ],
        "type": "ModelReference"
      },
```

**Comment by BirgitBoss** [comment]
Perhaps to add a third way of building paths to model elements. In Part 2 API of the AAS Specification there are so-called "idShort-Paths" defined (see Chapter 12.4 in V3.0 Addressing Resources).
For the example `(Submodel)https://example.com/aas/1/1/1234859590, (SubmodelElementList)Documents, (SubmodelElementCollection)0, (MultiLanguageProperty)Title`

we would have the idShort-Path
`Documents[0].Title`

Using it in an http/REST call needs encoding leads to
`GET /submodels/<encoded )https://example.com/aas/1/1/1234859590>/submodel/submodelElements/<encoded idShort-Path>`

**Comment by BirgitBoss** [comment]
> The attributes themselves aren't referable but the content they hold is. All have a range of `SubmodelElement` or `Property` respectively. So when adding a `SubmodelElement` INTO these attributes, they are referable. However, they can't be referred to with the existing `Reference` mechanism. That's how they differ from - say - `AssetAdministrationShell.administration` whose target (`AdministrativeInformation`) doesn't inherit from `Referable`. But the elements in the mentioned list do hold `Referable`s, just like `SubmodelElementCollection.value` does.
> 
> I don't have a use-case which is why I'll let this topic rest. The `Reference` mechanism is underspecified in this instance.

see https://github.com/admin-shell-io/aas-specs/issues/346

**Comment by StenGruener** [comment]
@BirgitBoss is this one closed with closed #103?

**Comment by BirgitBoss** [comment]
Yes
